Add more layer effects

Wend1go opened the pull request:
- [x] Add filters present in PIXI.filters
- [ ] Add filters present in external repo https://github.com/pixijs/pixi-filters
- [x] Add Action/Condition to enable/disable filters

Some effects like the displacement filter need a sprite to work, so I left it out for now. But it would be very cool to use an existing image ressource for that.

The pixi-filters has a bunch of nice effects but it is located in a different repository. Since it needs to be ES5 complient, how would I have to add the external JS file without using the `import` command? 

The pixi filters all have a property `enabled` that could theoretically be set the same way as the other filter parameters but I think it would be cleaner to create a separate action and condition for that.
I just don't know yet in which file I need to add that condition/action to appear in the events list.

4ian commented:
> The pixi filters all have a property enabled that could theoretically be set the same way as the other filter parameters but I think it would be cleaner to create a separate action and condition for that.

Yes I think a separate action would be more interesting.

> The pixi-filters has a bunch of nice effects but it is located in a different repository. Since it needs to be ES5 complient, how would I have to add the external JS file without using the import command?

We're starting to reach a point where we would have a lot of potential effects, which could end up taking a "large" amount of space on disk. To keep exported games as lean as possible, I would rather think about a system to have extensions to be able to declare new effects :) 
Right now as you've seen they are all hardcoded in the game engine and hardcoded in the IDE.
The idea would be to:
* Add an API to Extension (in C++ and JS) to declare effects and the JS file to include.
* Allow the IDE to read these effects declared in extensions.
* Allow the code generation to only include the effects used in layers (or in the future, objects :))
* Allow the JS game engine to use these effects (probably like done for objects and behaviors, by scanning things inside the `gdjs` "namespace", searching for a special property indicating the presence of an effect).

I can see that all filters from Pixi filters (https://github.com/pixijs/pixi-filters/tree/master/filters) are in separate files, so that would make sense to have them in an extension but with multiple include files - so that only the filters actually used are included in the exported game :)

I'm fine for now going ahead with the effects bundled with Pixi.js - because they won't be impacting the size of the game files as already included.
To add more, like the one from pixi-filters, I think it would be interesting to see if we can add them using this extension system like behaviors/objects.

4ian commented on GDJS/Runtime/pixi-renderers/pixi-filters-tools.js:
```diff
@@ -96,6 +96,43 @@ gdjs.PixiFiltersTools._filters = {
+            return blur;
+        },
+        updateParameter: function(filter, parameterName, value) {
+            if (parameterName !== 'blur' &&
```
While it's not done for existing effects, I think it would be worth adding a bit of extra check to avoid invalid values to be set for the parameters. What do you think? For example, return if the kernelSize is not a valid value, or value is out of range, etc.

Wend1go replied:
Yep makes sense.
I have already implemented a check for the kernelSize in [EffectDescription.js](https://github.com/4ian/GDevelop/pull/1206/commits/c1da506b419e862c4fcbfdca0693b9d3629df41e#diff-6c71ccf608de8a3035773caafdc2230aR15) 
Moved the checks to `pixi-filters-tools.js` so that we don't have to declare the same check functions twice.

4ian commented:
Almost forgot to say it: can't wait for these sweet shiny new shaders to be in GDevelop 🤗Will unlock a tons of cool visual effects.

Wend1go commented:
Distributing filters in a modular way via Extension sounds good. Since my hollidays are approaching, I won't be able to work on this during the upcoming weeks. So take you time.
I'll try to add the enable/disable functionality at the weekend, feel free to merge this after that if you intend to do a new release. I can open a new PR for the additional effects later.

4ian commented on newIDE/app/src/EffectsList/EffectDescription.js:
```diff
@@ -11,8 +11,6 @@ export type EffectDescription = {|
   parameterDefaultValues: Array<EffectDefaultValue>,
 |};
 
-const clampValue = value => Math.max(0, Math.min(1, value));
```
Actually, both the editor and the game engine carry the responsibility of validating the user input, but for two different reasons:
- The game engine because we don't want to pass invalid values to the filters - that could make strange results or even things like crashs (in other words, no bugs).
- The editor because it's a bad user experience for the user to be able to enter invalid values - if we know about the valid values we should as possible avoid to store them.

It's of course entirely possible for the user to enter invalid values using events, but that's why we have the checks in the game engine. The second checks are the editor best effort to avoid you entering non sense things :)

Wend1go replied:
So should we keep a second copy of the validation functions for the editor?

4ian replied:
Yes let's keep a copy of it, one in the game engine and one in the editor. Will generally in programming, it's best to factor things, here it's better to have both of them because there serves a slightly different purpose and it's not harmful. Better have a copy rather than a bad abstraction :)

> Is it possible to check for valid values while entering them in the textbox?

Not really, the text fields are made so that changes are applied only when they are losing the focus. Doing validation during the writing is trickier and can lead to strange results (like, you enter an invalid value like 4, and it's getting correct immediately => people can think of a bug).

4ian commented on GDJS/Runtime/pixi-renderers/pixi-filters-tools.js:
```diff
@@ -1,5 +1,8 @@
 gdjs.PixiFiltersTools = function() {};
 
+const clampValue = (value, min, max) => Math.max(min, Math.min(max, value));
```
We're in the game engine, so ES5 only ;)
`var clampValue = function(value, min, max) { return Math.max(min, Math.min(max, value)); }`

Also, remember that classical JavaScript is a pain when it comes to global variable and variable scope. Using var to declare a variable outside of any scope like this will create a global variable, both polluting the global namespace and risking our function to be erased by something else. We don't want to run this risk, so instead be sure to always declare things inside functions (in which case they are function scoped) or attach them as "static" members:

`gdjs.PixiFiltersTools.clampValue = function(value, min, max) { return Math.max(min, Math.min(max, value)); }`;

then use it as `filter.uniforms[parameterName] = gdjs.PixiFiltersTools.clampValue(value, 0, 1);`

More verbose, but that's the price of a nice "self contained" code that won't break even if someone mess with the global namespace.

4ian replied:
(note that the "someone mess with the global namespace" could be actually another extension declaring a `clampValue` with a different implementation - imagine the strange bug and how hard to debug it is "my game effects are broken when I use this unrelated extension")

4ian replied:
Wrote about this here: https://github.com/4ian/GDevelop/commit/5c648e3f2b8c444bd99221cea56b92c00cdddddd

4ian commented on GDJS/Runtime/pixi-renderers/pixi-filters-tools.js:
```diff
@@ -1,5 +1,8 @@
 gdjs.PixiFiltersTools = function() {};
 
+const clampValue = (value, min, max) => Math.max(min, Math.min(max, value));
+const clampKernelSize = value => ([5, 7, 9, 11, 13, 15].includes(value)) ? value : 5;
```
Same for this function. 
`includes` is also fairly recent and could create issue with old browser that we still support (read: Android 4.x). Use indexOf instead: `[5, 7, 9, 11, 13, 15].indexOf(value) !== -1`

4ian replied:
Note that in newIDE, we're running with Babel, so all the latest features of JS can be used, and using webpack to bundle JS, so declare function/variables anywhere is fine because they will be scoped to the JS file - no risk of collision.

4ian commented:
> I'll try to add the enable/disable functionality at the weekend, feel free to merge this after that if you intend to do a new release. I can open a new PR for the additional effects later.

Yep, both sounds good :)
- Check my comments, but apart from this I'm fine merging this as it's adding new filters without any price in terms of JS size, as they are already embedded in Pixi and so the exported games.
- Enable/disable will be a "bonus" if you have time, otherwise again it's already ok to go with this. You can even start another PR to avoid polluting this one and be sure I can merge it.
- I'll work on enabling extensions to provide effects next week or so.

Wend1go commented:
Ready for a final review @4ian

4ian commented on GDJS/Runtime/events-tools/cameratools.js:
```diff
@@ -128,6 +128,32 @@ gdjs.evtTools.camera.setLayerEffectParameter = function(runtimeScene, layer, eff
+ * @return {boolean} Filter is enabled
+ */
+gdjs.evtTools.camera.layerEffectEnabled = function(runtimeScene, layer, effect) {
+    if ( !runtimeScene.hasLayer(layer) ) { return 1; }
```
`return true`

4ian commented on GDJS/Runtime/layer.js:
```diff
@@ -216,6 +216,24 @@ gdjs.Layer.prototype.setEffectParameter = function(name, parameterIndex, value)
+};
+
+/**
+ * Enable an effect.
```
"Check if an effect is enabled"

4ian commented on GDJS/Runtime/layer.js:
```diff
@@ -216,6 +216,24 @@ gdjs.Layer.prototype.setEffectParameter = function(name, parameterIndex, value)
 };
 
+/**
+ * Enable an effect.
```
"Enable or disable an effect."

4ian commented on GDJS/Runtime/layer.js:
```diff
@@ -216,6 +216,24 @@ gdjs.Layer.prototype.setEffectParameter = function(name, parameterIndex, value)
 
+/**
+ * Enable an effect.
+ * @param {string} filter The name of the filter effect
```
"The name of the effect to enable or disable."

The term "filter" is Pixi.js specific, in GDevelop I think it's better to keep the metaphor of the "effect" everywhere.

4ian commented on GDJS/Runtime/pixi-renderers/layer-pixi-renderer.js:
```diff
@@ -110,3 +110,17 @@ gdjs.LayerPixiRenderer.prototype.setEffectParameter = function (name, parameterN
+};
+
+gdjs.LayerPixiRenderer.prototype.isEffectEnabled = function (name) {
+    if (!this._filters.hasOwnProperty(name)) return 1;
```
`return false` - an effect that does not exists must be considered as not enabled.

4ian commented on GDJS/Runtime/pixi-renderers/pixi-filters-tools.js:
```diff
@@ -93,11 +96,70 @@ gdjs.PixiFiltersTools._filters = {
 
+/**
+ * Enable an effect.
+ * @param {string} filter The name of the filter effect
```
This annotation is not right, it should ba a `{PIXI.Filter}` I think and the description should also be updated (`filter The filter to enable or disabled`)

4ian commented on GDJS/Runtime/pixi-renderers/pixi-filters-tools.js:
```diff
@@ -93,11 +96,70 @@ gdjs.PixiFiltersTools._filters = {
+}
+
+/**
+ * Enable an effect.
```
Same comments about description, filter parameter typing and description

4ian commented on GDJS/Runtime/pixi-renderers/pixi-filters-tools.js:
```diff
@@ -93,11 +96,70 @@ gdjs.PixiFiltersTools._filters = {
+/**
+ * Enable an effect.
+ * @param {string} filter The name of the filter effect
+ * @return {boolean} Filter is enabled
```
I'm usually more explicit about the values: `true` if the filter is enabled

Bouh commented:
Field value can be empty and crash preview
![image](https://user-images.githubusercontent.com/1670670/63657193-e1e2ad00-c79e-11e9-951c-c6541cd98dff.png)

Wend1go commented:
Ups, I forgot to update the function calls to use the min/max parameters when reverting the changes. 🤦‍♂ 
Will correct this tonight.

Wend1go commented:
Ok, fixed now.
I added a brightness filter as a little present, to make amends. ;)
This one should be useful to crate screen transitions with a fade to/from black effect.

Since all the filters only work on the WebGL renderer, I think we should add a condition to check if WebGL is available so one can react to this and add a workaround if filters can't be used.
This could be achieved by checking the type of renderer (canvas / webgl)  that was automatically created when the game started.

4ian commented:
Sweet! Code looks good, I'll take another look and merge this tonight :)

> (Btw: What is the reason for only allowing a single effect per layer?)

If I remember properly in GD4 when starting to implement this, the native game engine based on SFML was only supporting a single shader on layers - so I decided to limit it for HTML5 game engines and find a solution for SFML before allowing more effects. Turns out that this kind of concern is not relevant anymore - maybe in the future we'll have a new native game engine or what not - but that's not a reason to keep the limitation.
I'll enable multiple effects after merging this (still need a bit of rework in the game engine)

> Field value can be empty and crash preview

I'll take a look

> At the bottom of the layer list we have just a icon +, in other list we have a little text like "Click to add a objet"

Right I started changes to have proper buttons in most lists, but still missing this one and variables. I'll take a look.

For the scrolling issues, it's true that it's not ideal. Would need to use ScrollView component I think to solve this (feel free to take a look, it's less of a priority for me right now).

4ian commented:
I've made a few last fixes. Merging now! Thanks a lot for working on this 👍👍

4ian commented:
I've not forgotten about adding effects from extensions - feel free to ping me about this again if I'm not starting this :) I've been busy working on upgrading the internal UI components to a new version (with denser lists, native selects, etc).

Bouh commented:
Indeed texture is needed for Displacement effect.
I think we can put a default texture in base64.

With the paramters torepeat (wrapmode) the texture.
It would also be great to be able to animate the positions and rotation of the texture. To have a moving effect on the screen.
There is little missing for this addition.

I managed to put the effect with the JSevent but I can't get it to move.

i think we can use file opener for import custom texture in this effect.
In any case texture should be a parameters too.

[displacementFilter.zip](https://github.com/4ian/GDevelop/files/3721596/displacementFilter.zip)
